clients/drcachesim/tracer/tracer.cpp

```diff
@@ -201,6 +201,87 @@ drmemtrace_custom_module_data(void * (*load_cb)(module_data_t *module),
         return DRMEMTRACE_ERROR;
 }
 
+/* sideline */
+static drvector_t buf_queue;
+static uint queue_size = 0;
+static uint queue_size_mask = 0;
+static volatile uint queue_front = 0;
+static volatile uint queue_tail = 0;
+static void *sideline_exit_event;
+struct buf_entry_t {
+    file_t  file;  // file descriptor, 0 means the end of profiling.
+    void   *data;  // output data pointer, NULL means the end of data for this file.
+    ssize_t size;  // output data size.
+};
+
+static inline void
+trace_buf_enqueue(file_t file, void *data, ssize_t size)
+{
+    bool done = false;
+    buf_entry_t *entry = (buf_entry_t *)dr_global_alloc(sizeof(*entry));
+    entry->file = file;
+    entry->data = data;
+    entry->size = size;
+    do {
+        /* quick racy check without locking */
+        while (((queue_front + 1) & queue_size_mask) == queue_tail) {
+            /* the queue is full, waiting */;
+            dr_thread_yield();
+        }
+        drvector_lock(&buf_queue);
+        if (((queue_front + 1) & queue_size_mask) != queue_tail) {
+            drvector_set_entry(&buf_queue, queue_front, entry);
+            queue_front = (queue_front + 1) & queue_size_mask;
+            done = true;
+        }
+        drvector_unlock(&buf_queue);
+    } while (!done);
+}
+
+static inline buf_entry_t *
+trace_buf_dequeue()
+{
+    buf_entry_t *entry = NULL;
+    drvector_lock(&buf_queue);
+    if (queue_tail != queue_front) {
+        entry = (buf_entry_t *)drvector_get_entry(&buf_queue, queue_tail);
+        queue_tail = (queue_tail + 1) & queue_size_mask;
+    }
+    drvector_unlock(&buf_queue);
+    return entry;
+}
+
+static void
+sideline_run(void *arg)
+{
+    buf_entry_t *entry;
+    bool profiling = true;
+    do {
+        /* quick racy check without locking */
+        while (queue_tail == queue_front) {
+            /* the queue is empty, waiting */
+            dr_thread_yield();
+        }
+        entry = trace_buf_dequeue();
+        if (entry == NULL)
+            continue;
+        if (entry->data != NULL) {
+            file_ops_func.write_file(entry->file, entry->data, entry->size);
+            dr_raw_mem_free(entry->data, max_buf_size);
+        } else if (entry->file != 0) {
+            /* XXX: we cannot simply close the file if we have more than one
+             * sideline thread, because there might be some sideline thread
+             * writing to this file.
+             */
+            file_ops_func.close_file(entry->file);
+        } else {
+            profiling = false;
```

Comment:
So at process exit you put a single file==0 entry in the queue -- but only one thread will grab that, and you want to exit all threads.  This doesn't seem like it will scale to more than one thread easily as a method of exiting.  This is one reason why being incremental by having just one thread may not be the best approach: we need to be sure the design works with a real pool.  IMHO it is not worth putting in this sentinel entry scheme if it will not work with multiple threads.